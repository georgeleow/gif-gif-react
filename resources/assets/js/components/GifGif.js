import React, { Component } from 'react';
import ReactDOM from 'react-dom';

class GifGif extends Component {
    render() {
        return (
            <div>
                <div className="bg-white rounded shadow text-grey-darker">
                    <div className="py-8">
                        <h3 className="text-center">How do you pronounce the word <span className="font-light uppercase">gif</span>?</h3>

                        <div className="flex justify-center py-8">
                            <label htmlFor="jif" className="mx-4">
                                GIF
                            </label>
                            <input type="radio" id="jif" value="jif" name="pronunciation" className="mr-4" />

                            <label htmlFor="gif" className="mx-4">
                                GIF
                            </label>
                            <input type="radio" id="gif" value="gif" name="pronunciation" />
                        </div>
                    </div>

                    <div className="flex justify-center">
                        <div className="w-3/4">
                            <input className="input"
                                   id="reason"
                                   type="text"
                                   placeholder="Why do you think it should be pronounced that way?" />
                        </div>
                    </div>

                    <div className="flex justify-center my-4">
                        <div className="w-3/4">
                            <input className="input"
                                   id="name"
                                   type="text"
                                   placeholder="What is your name?" />
                        </div>
                    </div>

                    <div className="flex justify-center">
                        <button id="submit"
                                onClick="add"
                                className="button">
                            Submit
                        </button>
                    </div>
                </div>

                <div className="py-8">
                    <h4 className="text-grey-darker text-center">User Responses:</h4>
                    <ul>
                        <li className="py-4">
                            <p><span className="font-light uppercase">gif</span> â€” </p>
                            <p className="text-sm mt-2">Submitted by <span className="uppercase"> It is the way the creator of the format wanted!</span></p>
                        </li>
                    </ul>
                </div>
            </div>
        );
    }
}

export default GifGif;

if (document.getElementById('gif-gif')) {
    ReactDOM.render(<GifGif />, document.getElementById('gif-gif'));
}
