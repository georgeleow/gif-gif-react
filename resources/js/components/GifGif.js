import React, { Component } from 'react';
import ReactDOM from 'react-dom';

class GifGif extends Component {
    constructor(props) {
        super(props);

        this.state = {
            picked: '',
            reason: '',
            name: '',
            votes: JSON.parse(props.votes),
        }
    }

    addVote() {
        const { picked, reason, name } = this.state;

        axios.post('/api/votes', { picked, reason, name })
            .then(() => {
                this.setState({
                    votes: this.state.votes.concat({ picked, reason, name }),
                    picked: '',
                    reason: '',
                    name: '',
                });
            });
    }

    isValid() {
        return ! this.state.picked || ! this.state.reason || ! this.state.name;
    }

    renderVotes() {
        return this.state.votes.map((vote, i) => {
            return (
                <li className="py-4" key={i}>
                    <p><span className="font-light uppercase">gif</span> â€” { vote.reason }</p>
                    <p className="text-sm mt-2">Submitted by <span className="uppercase">{ vote.name }</span></p>
                </li>
            );
        })
    }

    render() {
        return (
            <div>
                <div className="bg-white rounded shadow text-grey-darker">
                    <div className="py-8">
                        <h3 className="text-center">How do you pronounce the word <span className="font-light uppercase">gif</span>?</h3>

                        <div className="flex justify-center py-8">
                            <label htmlFor="jif" className="mx-4">
                                GIF
                            </label>

                            <input type="radio" id="jif"
                                   onClick={() =>{ this.setState({ picked: 'jif' })}}
                                   checked={ this.state.picked === 'jif' }
                                   className="mr-4" />

                            <label htmlFor="gif" className="mx-4">
                                GIF
                            </label>

                            <input type="radio"
                                   id="gif"
                                   onClick={() =>{ this.setState({ picked: 'gif' })}}
                                   checked={ this.state.picked === 'gif' } />
                        </div>
                    </div>

                    <div className="flex justify-center">
                        <div className="w-3/4">
                            <input className="input"
                                   id="reason"
                                   type="text"
                                   value={ this.state.reason }
                                   onChange={ (e) => { this.setState({ reason: e.target.value })}}
                                   placeholder="Why do you think it should be pronounced that way?" />
                        </div>
                    </div>

                    <div className="flex justify-center my-4">
                        <div className="w-3/4">
                            <input className="input"
                                   id="name"
                                   type="text"
                                   value={ this.state.name }
                                   onChange={ (e) => { this.setState({ name: e.target.value })}}
                                   placeholder="What is your name?" />
                        </div>
                    </div>

                    <div className="flex justify-center">
                        <button id="submit"
                                onClick={() => { this.addVote() }}
                                disabled={ this.isValid() }
                                className={ `button ${ this.isValid() ? 'opacity-50 cursor-not-allowed' : ''}`}>
                            Submit
                        </button>
                    </div>
                </div>

                <div className="py-8">
                    <h4 className="text-grey-darker text-center">User Responses:</h4>
                    <ul>
                        { this.renderVotes() }
                    </ul>
                </div>
            </div>
        );
    }
}

export default GifGif;

let el;

if (el = document.getElementById('gif-gif')) {
    ReactDOM.render(<GifGif { ...el.dataset} />, el);
}
