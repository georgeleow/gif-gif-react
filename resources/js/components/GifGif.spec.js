import React from 'react';
import { mount } from 'enzyme';
import GifGif from './GifGif';

window.axios = {
    post: jest.fn().mockImplementation(() => Promise.resolve([]))
};

describe('GifGif', () => {
    let wrapper;

    beforeEach(() => {
        wrapper = mount(<GifGif votes={JSON.stringify([])} />);
    });

    it('renders', () => {
        expect(wrapper.text()).toContain('How do you pronounce the word gif?');
        expect(wrapper).toMatchSnapshot();
    });

    it('renders with props', () => {
        let vote = {
            picked: 'jif',
            reason: 'It is how you pronounce gin',
            name: 'Samantha'
        };

        wrapper = mount(<GifGif votes={ JSON.stringify([vote]) } /> );

        expect(wrapper.text()).toContain(vote.reason);
        expect(wrapper.text()).toContain(vote.name);
        expect(wrapper).toMatchSnapshot();
    });

    it('updates the selected pronuncation', () => {
        wrapper.find('#jif').simulate('click');
        expect(wrapper.state('picked')).toBe('jif');

        wrapper.find('#gif').simulate('click');
        expect(wrapper.state('picked')).toBe('gif');
    });

    it('submits a vote', async () => {
        let vote = {
            picked: 'jif',
            reason: 'It is how you pronounce gin',
            name: 'Samantha'
        };

        expect(wrapper.state('votes').length).toBe(0);

        wrapper.find('#jif').simulate('click');
        expect(wrapper.state('picked')).toBe('jif');

        // Update the reason input
        let reason = wrapper.find('#reason');
        reason.simulate('change', { target: { value: vote.reason }});
        expect(wrapper.state('reason')).toBe(vote.reason);

        // Update the name input
        let name = wrapper.find('#name');
        name.simulate('change', { target: { value: vote.name }});
        expect(wrapper.state('name')).toBe(vote.name);

        // Click the submit button
        wrapper.find('#submit').simulate('click');

        // Look, mom! No wrapper.vm.$nextTick() !
        await expect(axios.post).toBeCalledWith('/api/votes', vote);

        // Check that vote is added to vote
        expect(wrapper.state('votes').length).toBe(1);

        // Check that the vote is rendered somewhere
        expect(wrapper.text()).toContain(vote.reason);
        expect(wrapper.text()).toContain(vote.name);
        expect(wrapper.state('reason')).not.toBe(vote.reason);
        expect(wrapper.state('name')).not.toBe(vote.name);
    });

    it('cannot submit without all fields', () => {
        wrapper.find('#submit').simulate('click');
        expect(wrapper.state('votes').length).toBe(0);
    });
});
