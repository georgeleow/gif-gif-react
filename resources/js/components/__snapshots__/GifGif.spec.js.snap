// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GifGif renders 1`] = `
<GifGif
  votes="[]"
>
  <div>
    <div
      className="bg-white rounded shadow text-grey-darker"
    >
      <div
        className="py-8"
      >
        <h3
          className="text-center"
        >
          How do you pronounce the word 
          <span
            className="font-light uppercase"
          >
            gif
          </span>
          ?
        </h3>
        <div
          className="flex justify-center py-8"
        >
          <label
            className="mx-4"
            htmlFor="jif"
          >
            GIF
          </label>
          <input
            checked={false}
            className="mr-4"
            id="jif"
            onClick={[Function]}
            type="radio"
          />
          <label
            className="mx-4"
            htmlFor="gif"
          >
            GIF
          </label>
          <input
            checked={false}
            id="gif"
            onClick={[Function]}
            type="radio"
          />
        </div>
      </div>
      <div
        className="flex justify-center"
      >
        <div
          className="w-3/4"
        >
          <input
            className="input"
            id="reason"
            onChange={[Function]}
            placeholder="Why do you think it should be pronounced that way?"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        className="flex justify-center my-4"
      >
        <div
          className="w-3/4"
        >
          <input
            className="input"
            id="name"
            onChange={[Function]}
            placeholder="What is your name?"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        className="flex justify-center"
      >
        <button
          className="button opacity-50 cursor-not-allowed"
          disabled={true}
          id="submit"
          onClick={[Function]}
        >
          Submit
        </button>
      </div>
    </div>
    <div
      className="py-8"
    >
      <h4
        className="text-grey-darker text-center"
      >
        User Responses:
      </h4>
      <ul />
    </div>
  </div>
</GifGif>
`;

exports[`GifGif renders with props 1`] = `
<GifGif
  votes="[{\\"picked\\":\\"jif\\",\\"reason\\":\\"It is how you pronounce gin\\",\\"name\\":\\"Samantha\\"}]"
>
  <div>
    <div
      className="bg-white rounded shadow text-grey-darker"
    >
      <div
        className="py-8"
      >
        <h3
          className="text-center"
        >
          How do you pronounce the word 
          <span
            className="font-light uppercase"
          >
            gif
          </span>
          ?
        </h3>
        <div
          className="flex justify-center py-8"
        >
          <label
            className="mx-4"
            htmlFor="jif"
          >
            GIF
          </label>
          <input
            checked={false}
            className="mr-4"
            id="jif"
            onClick={[Function]}
            type="radio"
          />
          <label
            className="mx-4"
            htmlFor="gif"
          >
            GIF
          </label>
          <input
            checked={false}
            id="gif"
            onClick={[Function]}
            type="radio"
          />
        </div>
      </div>
      <div
        className="flex justify-center"
      >
        <div
          className="w-3/4"
        >
          <input
            className="input"
            id="reason"
            onChange={[Function]}
            placeholder="Why do you think it should be pronounced that way?"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        className="flex justify-center my-4"
      >
        <div
          className="w-3/4"
        >
          <input
            className="input"
            id="name"
            onChange={[Function]}
            placeholder="What is your name?"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        className="flex justify-center"
      >
        <button
          className="button opacity-50 cursor-not-allowed"
          disabled={true}
          id="submit"
          onClick={[Function]}
        >
          Submit
        </button>
      </div>
    </div>
    <div
      className="py-8"
    >
      <h4
        className="text-grey-darker text-center"
      >
        User Responses:
      </h4>
      <ul>
        <li
          className="py-4"
          key="0"
        >
          <p>
            <span
              className="font-light uppercase"
            >
              gif
            </span>
             â€” 
            It is how you pronounce gin
          </p>
          <p
            className="text-sm mt-2"
          >
            Submitted by 
            <span
              className="uppercase"
            >
              Samantha
            </span>
          </p>
        </li>
      </ul>
    </div>
  </div>
</GifGif>
`;
